core/chain_util.go

```diff
@@ -52,8 +53,16 @@ func CalcDifficulty(time, parentTime uint64, parentDiff *big.Int) *big.Int {
 		diff.Sub(parentDiff, adjust)
 	}
 	if diff.Cmp(params.MinimumDifficulty) < 0 {
-		return params.MinimumDifficulty
+		diff = params.MinimumDifficulty
 	}
+
+	periodCount := new(big.Int).Add(parentNumber, big.NewInt(1))
+	periodCount.Div(periodCount, expDiffPeriod)
+	if periodCount.Cmp(big.NewInt(1)) > 0 {
+		// diff = diff + 2^(periodCount - 2)
+		diff = common.BigMax(diff.Add(diff, new(big.Int).Exp(big.NewInt(2), new(big.Int).Sub(periodCount, big.NewInt(2)), big.NewInt(0))), params.MinimumDifficulty)
+	}
```

Comment:
I personally find these calculations get more readable (and possibly more efficient) in
"SSA style" with one operation at a time.

```go
periodCount.Sub(periodCount, common.Big2)
periodCount.Exp(common.Big2, periodCount, nil)
diff.Add(diff, periodCount)
diff = common.BigMax(diff, params.MinimumDifficulty)
```